scapy/layers/usb.py

```diff
@@ -77,6 +72,94 @@
 }
 
 
+_requests = {
+    0x00: "GET_STATUS",
+    0x01: "CLEAR_FEATURE",
+    0x03: "SET_FEATURE",
+    0x05: "SET_ADDRESS",
+    0x06: "GET_DESCRIPTOR",
+    0x07: "SET_DESCRIPTOR",
+    0x08: "GET_CONFIGURATION",
+    0x09: "SET_CONFIGURATION",
+    0x0A: "GET_INTERFACE",
+    0x0B: "SET_INTERFACE",
+    0x0C: "SYNCH_FRAME",
+}
+
+_request_recipients = {
+    0x00: "Device",
+    0x01: "Interface",
+    0x02: "Endpoint",
+    0x03: "Other",
+}
+
+_request_types = {
+    0x00: "Standard",
+    0x01: "Class",
+    0x02: "Vendor",
+    0x03: "Reserved",
+}
+
+_request_directions = {
+    0: "Device-to-host",
+    1: "Host-to-device",
+}
+
+_directions = {
+    0: "OUT",
+    1: "IN",
+}
+
+_urb_types = {
+    ord("S"): "SUBMIT",
+    ord("C"): "COMPLETE",
+    ord("E"): "ERROR",
+}
+
+
+class _Setup(Packet):
+    fields_desc = [BitEnumField("bmRequestType_direction", 0, 1, _request_directions),
+                   BitEnumField("bmRequestType_type", 0, 2, _request_types),
+                   BitEnumField("bmRequestType_recipient", 0, 5, _request_recipients),
+                   ByteEnumField("bRequest", 0, _requests),
+                   LEShortField("wValue", 0),
+                   LEShortField("wIndex", 0),
+                   LEShortField("wLength", 0)]
+
+
+class USBLinux(Packet):
+    name = "USBLinux URB"
+    fields_desc = [XLELongField("urb_id", 0),
+                   ByteEnumField("urb_type", 0, _urb_types),
+                   ByteEnumField("transfer_type", 0, _transfer_types),
+                   BitEnumField("direction", 0, 1, _directions),
+                   BitField("endpoint", 0, 7),
+                   ByteField("device_address", 0),
+                   LEShortField("bus_id", 0),
+                   StrFixedLenField("setup_flag", b"\x00", length=1),
+                   StrFixedLenField("data_flag", b"\x00", length=1),
+                   LELongField("urb_ts_sec", 0),
+                   LESignedIntField("urb_ts_usec", 0),
+                   LESignedIntField("urb_status", 0),
+                   LEIntField("urb_len", 0),
+                   LEIntField("urb_data_len", 0),
+                   _Setup]
+
+
+class _USBLinuxExt(Packet):
+    name = "USBLinux URB"
+    fields_desc = [LESignedIntField("interval", 0),
+                   LESignedIntField("start_frame", 0),
+                   LEIntField("xfer_flags", 0),
+                   LEIntField("numdesc", 0)]
```

Comment:
I wasn't sure about how to handle these extra fields. For example, if I wanted to create a `SetFeatureRequest` `Packet` subclass, I would need two versions of it: one for the 48 byte header and one for the 64 byte header. So I broke this out into `_USBLinuxExt` so we can reuse it for making these different variants of packets.

I was hoping there was a cleaner way to do it, but I couldn't figure anything out. Let me know if there's a better way to do this.

Reply:
You could just add those fields at the end of the extended header (instead of putting them in a separate packet), and keep the first field of the extended packet as the non-extended packet. (in which case its fields are copied)